---
title: "Spearman on enterobacteriacea only"
output: html_notebook
author: Zakhar Krekhno
---

```{r, echo=FALSE, include=FALSE}
library(readxl)
library(reshape2)
library(psych)
library(tidyverse)
library(ggprism)
library(rstatix)
theme_set(theme_classic(base_size = 14))
```

Import all the necessary matrices for analysis
```{r import}
files <- list.files(path = here::here("corr_mats/"))
otu_file_list <- str_subset(files,"clr(.)+otu")
tax_file_list <- str_subset(files,"tax")
ms_file_list <- str_subset(files, "ms")
#Import otu counts
otu_list <- map(otu_file_list, 
                    ~read_csv(file = here::here(str_c("corr_mats/",.))))
otu_mat_list <- otu_list %>% 
  map(column_to_rownames, var = "...1") %>% 
  map(as.matrix)
#Import MS data
mq_mat <- read_csv(file = here::here(str_c("corr_mats/",ms_file_list))) %>% 
  column_to_rownames(var = "...1") %>% 
  as.matrix()
```

Import the metadata and create sub-divided matrices.
We will be doing correlations within just Ileum or Colon and within noIBD, active, or inactive IBD
```{r import-meta}
meta_full <- read_excel(here::here("Metadata.xlsx")) %>% 
  select("sample" = `#SampleID`, "group" = Group, everything()) %>% 
  mutate(sample = str_remove(sample, "Rep")) %>% 
  filter(sample %in% rownames(mq_mat)) %>% 
  mutate(across(matches("(Active)|(TI_resection)|(Cholestyramine)"), ~as.character(fct_recode(as_factor(.x), 
                                                Yes = "1", No = "0", Not_applicable = "NA"))))
#Prepare metadata tables
meta_list_starter <- map(vector(mode = "list", length = 6), ~ meta_full)
meta_list <-meta_list_starter %>% 
  map2(rep(c("Colon","Ileum"), each = 3), 
       ~filter(.x, Tissue_type == .y)) %>% 
  map2(rep(c("Not_applicable","No","Yes"),times = 2), 
       ~filter(.x, Active_IBD == .y))
#Name the list for convenience
subgroups <- str_c(rep(c("Colon","Ileum"), each = 3),rep(c("Not_applicable","No","Yes"),times = 2), sep = "_")
names(meta_list) <- subgroups
```

Make appropriate matrices for each group
```{r subset_mq}
dfs_names <- str_c(rep(c("Family","Genus", "Species"), times = 6),rep(subgroups,each = 3), sep = "_")
otu_filt_list <- rep(otu_mat_list, times = 6) %>% 
  map2(rep(meta_list, each=3), ~ .x [pull(.y,var = sample),]) %>% 
  set_names(nm=dfs_names)
```

Further filter the proteomics data to just include DE proteins between active/inactive IBD vs control in each group
```{r select_top_prots}
#Import the results
da_res <- map(list.files(here::here("spear_subset_on_Foster"), pattern = "full_"),
              ~read_csv( file = here::here(str_c("spear_subset_on_Foster/",.)))) 
da_res <- set_names(da_res, nm = list.files(here::here("spear_subset_on_Foster"), pattern = "full_"))
#Reduce the results to one table, and filter afterwards
da_res_interest <- da_res %>% 
  reduce(left_join, by = c("gene.name","first.id")) %>% 
  select(gene.name,first.id, contains("padj"), -contains("grand")) %>% 
  filter(if_any(contains("adj"), ~.<0.01))
#Get the protein names to select
da_prots <- da_res_interest %>% 
  select(gene.name) %>% 
  separate(gene.name, into = letters [1:20], sep = " ") %>% 
  pivot_longer(letters [2:20], names_to = "letters", values_to = "secondary_names")
da_prots_final <- c(da_prots$a,da_prots$secondary_names) %>% 
  . [!is.na(.)] %>% 
  str_remove_all(";")

#Create the new mq_list
mq_top_list <- map(vector(mode = "list", length = 6*3), ~ mq_mat [,colnames(mq_mat) %in% da_prots_final]) %>% 
  map2(rep(meta_list, each=3), ~ .x [pull(.y,var = sample),]) %>% 
  set_names(nm=dfs_names)
#Ensure rows are in the same order for the matrices
all(map2_lgl(otu_filt_list, mq_top_list, ~all(rownames(.x) == rownames(.y))))
```

Filter OTUs to only include Enterobacteriacea

```{r}
#Import tax table
tax <- read_csv(here::here("corr_mats/Species_tax.csv"))
#Check that all features have taxonomy
otu_filt_list %>% 
  map_lgl(~all(colnames(.x) %in% tax$...1)) %>% 
  all()
#Filter the otu matrices to only include Entero features
otu_entero_list <- otu_filt_list %>% 
  map(~. [,colnames(.) %in% pull(filter(tax, Family == "f__Enterobacteriaceae"), var = "...1")])
entero_features <- otu_filt_list %>% 
  map(~ subset(colnames(.),colnames(.) %in% pull(filter(tax, Family == "f__Enterobacteriaceae"), var = "...1")))
otu_entero_final <- keep(otu_entero_list, str_detect(names(otu_entero_list), "Species", negate = T))
mq_top_final <- keep(mq_top_list, str_detect(names(mq_top_list), "Species", negate = T))
#Ensure rows are in the same order for the matrices
all(map2_lgl(otu_entero_final, mq_top_final, ~all(rownames(.x) == rownames(.y))))
```


Run correlations between Entero and 4 bile acid proteins: ASBT,   I-BABP, PTPN6,    SLC51A in Colon (IBD and non-IBD), and IBD colon, resected or non-resected TI
    
```{r entero-bile, fig.width=8,fig.height=6}
#Prepare metadata tables for bile acids
meta_list_ibd_ti <-meta_list_starter [1:2] %>% 
  map2(c("Control","IBD"),
       ~filter(.x, group == .y)) 

meta_list_ibd_ti [3:4] <- map(vector(mode = "list", length = 2), ~ meta_list_ibd_ti [[2]]) %>% 
  map2(c("No","Yes"), ~filter(.x, TI_resection == .y))

names(meta_list_ibd_ti) <- c("Control","IBD", "IBD_with_TI","IBD_no_TI")
#Select proteins of interest, names are synonyms
mq_bile_prot <- mq_mat [, c("SLC10A2","FABP6","PTPN6","SLC51A")]
mq_bile_list <- meta_list_ibd_ti %>% 
  map(~ mq_bile_prot [pull(filter(.x, Tissue_type == "Ileum"),var=sample),])
#Prepare otu counts matrices
entero_bile_list <- meta_list_ibd_ti %>% 
  map(~otu_mat_list [[1]] [pull(filter(.x, Tissue_type == "Colon"),var=sample), 
                     colnames(otu_mat_list [[1]]) %in% pull(filter(tax, Family == "f__Enterobacteriaceae"), var = "...1")])
#Remove unmatching samples
entero_samples_to_keep <- map2(entero_bile_list, mq_bile_list,
                               ~intersect(names(.x),str_replace(rownames(.y),"I$","C")))
mq_samples_to_keep <- map2(entero_bile_list, mq_bile_list,
                           ~intersect(str_replace(names(.x),"C$","I"), rownames(.y)))
entero_bile_match_list <- map2(entero_bile_list, entero_samples_to_keep,
                               ~ .x [.y])
mq_bile_match_list <- map2(mq_bile_list, mq_samples_to_keep,
                               ~ .x [.y,])

#Confirm that things are in proper order
map2_lgl(entero_bile_match_list, mq_bile_match_list, ~all(names(.x) == str_replace(rownames(.y),"I$","C")))

tidy_bile_list <- map2(mq_bile_match_list,entero_bile_match_list,
                       ~cbind(.x,.y)) %>%
  map(as.data.frame)
dummy_vars <- c("SLC10A2","FABP6","PTPN6","SLC51A")
#Do the correlation analysis
spear_bile_list <- map(tidy_bile_list, 
                       ~cor_test(data = .x, 
                                 vars = all_of(dummy_vars), 
                                 vars2 = ".y",
                                 method = "spearman")) %>% 
  map(mutate, protein = var1) 

tidy_bile_graphing <- tidy_bile_list %>% 
  map(pivot_longer, cols = -.y, names_to = "protein", values_to = "protein_abundance") %>% 
  map2(spear_bile_list, left_join, by = c("protein")) %>% 
  map(group_by,protein) %>% 
  map(mutate, text_y = max(protein_abundance)+1)

#pdf(here::here("spear_entero/Entero_bile_proteins_cors.pdf"), width = 8, height = 6)
walk2(tidy_bile_graphing, names(tidy_bile_graphing),
      ~print(ggplot(.x, aes(x=.y, y = protein_abundance)) +
               geom_smooth(method = "lm", lwd =1.2, color = "brown") +
               geom_point() + 
               geom_text(aes(x=0,y=text_y, label = str_c("rho = ",cor, "; p = ",round(p,2))), size = 4) +
               facet_wrap(~protein, nrow = 2, scales = "free_y") +
               theme_prism(base_size = 14) +
               labs(caption = .y, x="family Entero Abundance in colon", y = "Protein abundance in ileum")))
#dev.off()

```




Run correlations on subset proteins

```{r spearman_top_da, warning=FALSE, eval=FALSE}
spear_top_list <- map2(mq_top_final,otu_entero_final,
                     ~corr.test(x = .x, y = .y, use = "pairwise", method="spearman", adjust="BH", ci=FALSE))
coef_top_list <- map2(rep(spear_top_list,each=3),rep(c("r","p","p.adj"),times = length(spear_top_list)), ~melt(.x [.y]))

rs_top_list <- coef_top_list [seq(from = 1, to = length(coef_top_list), by = 3)]
ps_top_list <- coef_top_list [seq(from = 2, to = length(coef_top_list), by = 3)]
padj_top_list <- coef_top_list [seq(from = 3, to = length(coef_top_list), by = 3)]

df_top_list <- pmap(list(rs_top_list,ps_top_list,padj_top_list),
               ~reduce(list(..1,..2,..3),left_join, by = c("Var1","Var2")))
dfs_names <- str_c(rep(c("Family","Genus"), times = 6),rep(subgroups,each = 2), sep = "_")
names(df_top_list) <- dfs_names
#Filter the results to only include somewhat meaningful work
df_top_filt_list <- df_top_list %>% 
  map(~filter(., value < 0.05))
df_top_cors <- df_top_filt_list %>% 
   map(~arrange(.,value,desc(abs(as.numeric(value.x))))) %>% 
  map(slice_head,n=1) %>% 
  keep(~nrow(.)>0) %>% 
  map(mutate, Var1 = as.character(Var1), Var2 = as.character(Var2))
#walk2(df_top_list, dfs_names,
 #     ~write_csv(.x,here::here(str_c("spear_entero/",.y,"_entero.csv"))))

```

Plot scatter plots of top correlations

```{r scatter-top}
#Select just protein and otu as selected top correlation
mq_select_list <- mq_top_final %>% 
  keep(names(.) %in% names(df_top_cors)) %>% 
  map2(df_top_cors, ~ .x [,colnames(.x) %in% pull(.y,var = "Var1")]) %>%  
  map(as.data.frame, stringsAsFactors = F)
otu_select_list <- otu_entero_final %>% 
  keep(names(.) %in% names(df_top_cors)) %>%  
  map(as.data.frame, stringsAsFactors = F)
#prepare list for scatter plots
scatter_list <- map2(mq_select_list, otu_select_list, merge, by = 0)
#prepare lists for names
xlabs <- map_chr(df_top_cors, pull, var = "Var1")
ylabs <- map(vector(mode = "list", length = length(otu_select_list)), ~ tax) %>% 
  map2(keep(entero_features, names(entero_features) %in% names(df_top_cors)),
       ~filter(.x, `...1` == .y)) %>% 
  map2(rep(c("Family","Genus"), times = 2), ~pull(.x, var = .y)) %>% 
  flatten_chr()
rcoefs <- map_chr(df_top_cors, pull, var = "value.x")
pvals <- map_chr(df_top_cors, pull, var = "value")

#pdf(here::here("spear_entero/Scatter plots of top correlations entero.pdf"))
pwalk(list(scatter_list, xlabs,ylabs, rcoefs,pvals, names(scatter_list)), 
      ~print(ggplot(..1, aes(x = `.x[[i]].x`, y = `.x[[i]].y`)) +
               geom_point() +
               geom_smooth(method = "lm") + 
               labs(x = ..2, y = ..3, 
                    subtitle = str_c("Spearman rho =",round(as.numeric(..4),2)," Adjusted p=",
                                     format(as.numeric(..5),digits = 2, scientific = T)), title = ..6)
             ))
  
#dev.off()
```
Plot bubble plots of Spearman correlations
```{r spearman-plot-bubble, fig.height=6, fig.width=8}
df_top_reduced <- keep(df_top_filt_list, ~nrow(.) > 0) %>% 
  map2(ylabs, ~mutate(.x, Var2 = .y))

#Get taxonomic ranks for plotting
ranks <- str_remove(names(df_top_reduced),"_.+")

#pdf(here::here("spear_entero/subset Entero Spearman correlations.pdf"), height = 6, width = 8)
pwalk(list(df_top_reduced,ranks, names(df_top_reduced)),
      ~print(ggplot(..1, aes_(x=quote(Var1), y=quote(Var2))) +
              geom_point(aes(color = value.x, fill = value.x, size = value), alpha = 0.6, shape = 21) +
              scale_fill_distiller(name = "Spearman rho",palette = "PuOr", breaks = c(-0.6,-0.4,-0.2,0,0.2,0.4,0.6)) +
              scale_color_distiller(name = "Spearman rho", palette = "PuOr",breaks = c(-0.6,-0.4,-0.2,0,0.2,0.4,0.6)) +
              scale_size(name = "Adjusted p",trans = 'reverse', range = c(5,8)) +
              labs(x=element_blank(), y=element_blank(),
                   title = ..3, subtitle = "Yes means active IBD, No means inactive,\nNot applicable is control") +
              theme_bw(base_size = 16) + theme(axis.text.x = element_text(angle = 45, hjust = 0.9))))
#dev.off()
```
