---
title: "Clustering proteins"
output: html_notebook
---

```{r, echo=FALSE, include=FALSE}
library(readxl)
library(reshape2)
library(psych)
library(phyloseq)
library(dendextend)
library(gplots)
library(clusterProfiler)
library(tidyverse)
library(ggprism)
theme_set(theme_prism(base_size = 14))
```

Import all the necessary matrices for analysis
```{r import}
files <- list.files(path = here::here("corr_mats/"))
ms_file_list <- str_subset(files, "ms")
#Import MS data
mq_mat <- read_csv(file = here::here(str_c("corr_mats/",ms_file_list))) %>% 
  column_to_rownames(var = "...1") %>% 
  as.matrix()
```
We will be doing correlations within just Ileum or Colon and within noIBD, active, or inactive IBD
```{r import-meta}
meta_full <- read_excel(here::here("Metadata.xlsx")) %>% 
  select("sample" = `#SampleID`, "group" = Group, everything()) %>% 
  mutate(sample = str_remove(sample, "Rep")) %>% 
  filter(sample %in% rownames(mq_mat)) %>% 
  mutate(across(matches("(Active)|(TI_resection)|(Cholestyramine)"), ~as.character(fct_recode(as_factor(.x), 
                                                Yes = "1", No = "0", Not_applicable = "NA"))))
#Prepare metadata tables
meta_group_list <- vector(mode = "list", length = 4) %>% 
  map(~ meta_full) %>%  
  map2(rep(c("Colon","Ileum"), each = 2), 
       ~filter(.x, Tissue_type == .y)) %>% 
  map2(rep(c("Control","IBD"),times = 2), 
       ~filter(.x, group == .y)) %>% 
  set_names(nm = str_c(rep(c("Colon","Ileum"), each = 2), rep(c("Control","IBD"),times = 2), sep = "_"))
meta_active_list <- vector(mode = "list", length = 6) %>% 
  map(~ meta_full) %>%  
  map2(rep(c("Colon","Ileum"), each = 3), 
       ~filter(.x, Tissue_type == .y)) %>% 
  map2(rep(c("Not_applicable","No","Yes"),times = 2), 
       ~filter(.x, Active_IBD == .y)) %>% 
  set_names(str_c(rep(c("Colon","Ileum"), each = 3),rep(c("Not_applicable","No","Yes"),times = 2), sep = "_"))
meta_list <- c(meta_group_list, meta_active_list)

#Keep list names for convenience
subgroups <- names(meta_list)
```
Further filter the proteomics data to just include DE proteins between active/inactive IBD vs control in each group
```{r select_top_prots}
#Import the results
da_res <- map(list.files(here::here("spear_subset_on_Foster"), pattern = "full_"),
              ~read_csv( file = here::here(str_c("spear_subset_on_Foster/",.)))) 
da_res <- set_names(da_res, nm = list.files(here::here("spear_subset_on_Foster"), pattern = "full_"))
#Reduce the results to one table, and filter afterwards
#Filter for significance and effect 
da_res_interest <- da_res %>% 
  map(filter, if_any(contains("adj"), ~.<0.01), if_any(contains("eff"), ~abs(.)>0.5)) %>% 
  reduce(full_join, by = c("gene.name","first.id")) %>% 
  select(gene.name,first.id, contains("padj"), -contains("grand"), contains("eff")) 


#Get the protein names to select
da_prots <- da_res_interest %>% 
  select(gene.name) %>% 
  separate(gene.name, into = letters [1:20], sep = " ") %>% 
  pivot_longer(letters [2:20], names_to = "letters", values_to = "secondary_names")
da_prots_final <- c(da_prots$a,da_prots$secondary_names) %>% 
  . [!is.na(.)] %>% 
  str_remove_all(";")

#Create the new mq_list
mq_top_list <- map(meta_list, ~ mq_mat [,colnames(mq_mat) %in% da_prots_final]) %>% 
  map2(meta_list, ~ .x [pull(.y,var = sample),]) 
```




Run correlations on the top proteins in each comparison - only look at IBD colon and ileum for now

```{r spearman}
mq_select_top <- mq_top_list [c(2,4)]
#Run the spearman correlations
psych_cor_list <- mq_select_top %>% 
  map(corr.test, method = "spearman")
#Extract the coef and p-value matrices
psych_cor_mats_list <- psych_cor_list %>% 
  map(function(x) map(c("r","p"), 
                        ~x [.x]))
#Melt the matrices
#Create the helper function to assign NA to lower triangle of the matrix
assign_lower_tri_na <- function(x) {
  x [lower.tri(x, diag = T)] <- NA
  return(x)
}
psych_dfs_list <- psych_cor_mats_list %>% 
  map_depth(1, flatten) %>%  
  map_depth(2, assign_lower_tri_na) %>%  
  map_depth(2, melt) %>% 
  map_depth(2, filter, !is.na(value)) %>% 
  map_depth(1, reduce, left_join, by = c("Var1","Var2")) %>% 
  map(rename, r = value.x, p = value.y)
```

Now that correlation is done, we can run the clustering analysis with Ward linkages. Roughly following (this)[https://bio723-class.github.io/Bio723-book/clustering-in-r.html#combining-heatmaps-and-dendrograms] for the clustering analysis. 

```{r clustering}
#Transform the correlation into Distances
  
psych_coef_list <- psych_cor_list %>% 
  map(~.x$r) 
  
psych_dist <- psych_coef_list %>% 
  map(~as.dist(1-.x))

#Use Ward linkages to cluster the correlation data

psych_tree <- psych_dist %>% 
  map(hclust, method = "ward.D2")

psych_dend <- psych_tree %>% 
  map(as.dendrogram)


map(psych_dend, nleaves)
map(psych_dend, nnodes)
#Try to cut the dendrogram into clusters
cluster_number <- 6
psy_clust <- psych_dend %>% 
  map(cutree, k=cluster_number)
#Check how many proteins in each cluster
map(psy_clust, table)

walk2(psych_dend,names(psych_dend), ~plot(color_branches(.x, k=cluster_number), leaflab = "none", xlab = .y ))
```

We can try visualizing the results in a heatmap of correlations

```{r k-medoid-cluster, fig.width=6, fig.height=6}
#Create clusters using k-medoids
psy_kmed <- psych_dist %>% 
  map(cluster::pam, cluster_number)
map(psy_kmed, ~table(.x$cluster))
psy_kmed_clusters <- psy_kmed %>% 
  map(~.x$cluster)
```


```{r heatmap, fig.width=6, fig.height=6}
mycolors <- rcartocolor::carto_pal(name = "Geyser")

color_sheme <- colorRampPalette(mycolors) (255)

#pdf(here::here("protein_cluster_correlation/DA_clustering_heatmaps.pdf"), width = 6, height = 6)
pwalk(list(psych_coef_list, psych_dend,names(psych_coef_list)), 
     ~heatmap.2(..1, Rowv = NULL, Colv =NULL, dendrogram = "none", 
                col = color_sheme, 
                trace = "none", density.info = "none", revC = T,
                key.title = "Color Key", key.xlab = "Spearman's R", sub = ..3, keysize = 1.5,
                cexRow = 0.5,cexCol = 0.5, main = "No clustering", labCol = F, labRow = F))

pwalk(list(psych_coef_list, psych_dend,names(psych_coef_list)), 
     ~heatmap.2(..1, Rowv = ladderize(..2), Colv =ladderize(..2), dendrogram = "none", 
                col = color_sheme, 
                trace = "none", density.info = "none", revC = T,
                key.title = "Color Key", key.xlab = "Spearman's R", sub = ..3, keysize = 1.5,
                 cexRow = 0.5,cexCol = 0.5, main = "Ward clustering",labCol = F, labRow = F))
psy_coef_kmed_list <- psych_coef_list %>% 
  map2(psy_kmed_clusters, ~ .x [order(.y),order(.y)])
psy_kmed_cluster_cols <- psy_kmed_clusters %>% 
  map(~.x [order(.x)]) %>% 
  map(as_tibble) %>% 
  map(mutate, color = rcartocolor::carto_pal(name = "Antique") [value])
  
pwalk(list(psy_coef_kmed_list,names(psych_coef_list), psy_kmed_cluster_cols), 
     ~heatmap.2(..1, Rowv = NULL, Colv =NULL, dendrogram = "none", 
                col = color_sheme, 
                trace = "none", density.info = "none", revC = T,
                key.title = "Color Key", key.xlab = "Spearman's R", sub = ..2, keysize = 1.5,
                cexRow = 0.5,cexCol = 0.5, main = "K-medoid clustering", RowSideColors = pull(..3,var = "color"),
                labCol = F, labRow = F))
#dev.off()
```
Perform functional analysis of the clusters
```{r cluster-profiler}
h <- read.gmt(here::here("h.all.v7.5.1.symbols.gmt"))
cp <- read.gmt(here::here("c2.cp.v7.5.1.symbols.gmt")) %>% 
  filter(str_detect(term, "REACT")) %>% 
  mutate(term = str_remove(term,"REACTOME_"))
universe <- colnames(mq_mat)
#Get lists of Ward clusters from clustering data
psy_ward_cluster_dfs <- psy_clust %>% 
  map(melt) %>% 
  map(rownames_to_column, var = "protein") 
psy_ward_clusters <- psy_ward_cluster_dfs %>% 
  map(nest, data = "protein") %>% 
  map(~ .x [[2]]) %>% 
  map_depth(2, flatten_chr) %>% 
  map_depth(1, set_names, nm = str_c("cluster", 1:6, sep = "_"))

#Perform ORA analysis using either Hallmark or Reactome pathways

h_list <- psy_ward_clusters %>% 
  map_depth(2, 
            ~enricher(.x, TERM2GENE = h, 
                      pvalueCutoff = 0.05, pAdjustMethod = "BH", universe = universe,
                      qvalueCutoff = 0.05))

cp_list <- psy_ward_clusters %>% 
  map_depth(2, 
            ~enricher(.x, TERM2GENE = cp, 
                      pvalueCutoff = 0.05, pAdjustMethod = "BH", universe = universe,
                      qvalueCutoff = 0.05))

#Filter to only keep succesful clusters

suc_h_list <- h_list %>% 
  map(~keep(.x,~nrow(.x) > 0))
flat_suc_h_list <- suc_h_list %>% 
  map2(names(suc_h_list), ~set_names(.x,nm=str_c(.y, names(.x), sep = "__"))) %>% 
  flatten() 
suc_cp_list <- cp_list %>% 
  map(~keep(.x,~nrow(.x) > 0))
flat_suc_cp_list <- suc_cp_list %>% 
  map2(names(suc_cp_list), ~set_names(.x,nm=str_c(.y, names(.x), sep = "__"))) %>% 
  flatten() 
```

Visualize the results of functional profiling

```{r cluster-profiler-vis}
#Visualize functional results
h_dot_list <- flat_suc_h_list %>% 
  map2(names(flat_suc_h_list),~dotplot(.x, showCategory = 30) +
        scale_color_gradientn(colours = mycolors [4:7]) +
        labs(caption = .y, subtitle = "Hallmark"))
cp_dot_list <- flat_suc_cp_list %>% 
  map2(names(flat_suc_cp_list),~dotplot(.x, showCategory = 20) +
        scale_color_gradientn(colours = mycolors [4:7]) +
        labs(caption = .y, subtitle = "Reactome") +
         theme(axis.text.y = element_text(size = 5)))
cp_dot_list
h_heat_list <- flat_suc_h_list %>% 
  map2(names(flat_suc_h_list),~heatplot(.x, showCategory = 30) +
        labs(caption = .y, subtitle = "Hallmark"))

cp_heat_list <- flat_suc_cp_list %>% 
  map2(names(flat_suc_cp_list),~heatplot(.x, showCategory = 20) +
        labs(caption = .y, subtitle = "Reactome") +
         theme(axis.text.y = element_text(size = 5)))

#Combine the results into lists for grid plotting
h_dot_heat_list <- map2(h_dot_list, h_heat_list, ~set_names(list(.x,.y), nm = c("dot","heat")))
cp_dot_heat_list <- map2(cp_dot_list, cp_heat_list, ~set_names(list(.x,.y), nm = c("dot","heat")))
```

Plot the results on a grid
```{r cluster-profiler-grid, fig.width=6, fig.height=12} 
#pdf(here::here("protein_cluster_correlation/Hallmark_enrichment.pdf"), width = 6, height = 8)
map(h_dot_heat_list,
  ~cowplot::plot_grid(plotlist = .x, nrow = 2))
#dev.off()
#pdf(here::here("protein_cluster_correlation/Reactome_enrichment.pdf"), width = 6, height = 12)
map(cp_dot_heat_list,
  ~cowplot::plot_grid(plotlist = .x, nrow = 2))
#dev.off()
```
Next step is to average expression of the proteins within a cluster (only keep clusters that had some functionality identified) and correlate that with bacterial expression

```{r average_cluster_expression}
mq_cluster_expression <- mq_select_top %>% 
  map(melt) %>% 
  map(rename, protein = Var2, sample = Var1, abundance = value) %>% 
  map2(psy_ward_cluster_dfs, left_join, by = "protein") %>% 
  map(mutate, value = str_c("cluster",value, sep = "_")) %>% 
  map(group_by, value, sample) %>% 
  map(summarize, cluster_mean = mean(abundance)) 

mq_relevant_cluster_expr <- pmap(list(mq_cluster_expression, suc_h_list, suc_cp_list),
                                 ~ filter(..1, value %in% union(names(..2), names(..3))))

mq_cluster_mat <- mq_relevant_cluster_expr %>% 
  map(pivot_wider, names_from = value, values_from = cluster_mean) %>% 
  map(column_to_rownames, var = "sample") %>% 
  map(as.matrix)
```

Next get bacterial families
Import all the necessary matrices for analysis
```{r import-otu}
files <- list.files(path = here::here("corr_mats/"))
otu_file_list <- str_subset(files,"clr(.)+otu")
tax_file_list <- str_subset(files,"tax")
#Import otu counts
otu_list <- map(otu_file_list, 
                    ~read_csv(file = here::here(str_c("corr_mats/",.))))
otu_mat_list <- otu_list %>% 
  map(column_to_rownames, var = "...1") %>% 
  map(as.matrix)
#Only keep family level otus
otu_family <- otu_mat_list [[1]]
#Prepare the OTU matrices for each comparison
otu_fam_list <- mq_cluster_mat %>% 
  map(~ otu_family [rownames(.x),])
```


```{r select-top-otu}
biom <-  import_biom("table-with-taxonomy.biom", parseFunction = parse_taxonomy_default)
tree <- read_tree_greengenes("tree.nwk")
metadata <- import_qiime_sample_data("metadata.txt")
physeq <- merge_phyloseq(biom, metadata, tree)
colnames(tax_table(physeq)) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
#Remove irrelevant samples
physeq <- subset_samples(physeq, !(sample_names(physeq) %in% c("C34C","I01I","I60I","I12I", "I44I")))
#Remove all mocks, and negative controls
physeq <- subset_samples(physeq, str_detect(X.SampleID, "(Mock)|(Neg)|(TC7)|(H2O)", negate = T))
sample_names(physeq) <- str_remove(sample_names(physeq),"Rep")

phy_fam <- tax_glom(physeq, taxrank = "Family")
phy_fam_rel <- transform_sample_counts(phy_fam, function(x) x/sum(x)) 

otu_fam_rel <- mq_select_top %>% 
  map(~otu_table(phy_fam_rel)) %>% 
  map2(mq_select_top, ~.x [,rownames(.y)])


#In each case select a top n number of families (mean abundance)
top_cut <- 20
otu_top_fam <- otu_fam_rel %>%
  map(melt) %>% 
  map(group_by, Var1) %>%   
  map(summarise, mean_abund = mean(value)) %>% 
  map(~arrange(.x,desc(mean_abund))) %>% 
  map(slice_head, n=top_cut)

otu_fam_list <- mq_cluster_mat %>% 
  map(~ otu_family [rownames(.x),])

otu_fam_list <- otu_fam_list %>% 
  map2(otu_top_fam, ~.x [,colnames(.x) %in% pull(.y,var = "Var1")])
```

Run Spearman correlations
```{r cluster-otu-spear} 
cluster_fam_list <- cross2(mq_cluster_mat, otu_fam_list)
cluster_spear <- map(cluster_fam_list,
                      ~corr.test(x=.x [[1]],y=.x [[2]], method = "spearman", adjust = "BH"))
cluster_spear_r <- map(cluster_spear, 
                          ~melt(.x$r))
cluster_spear_p <- map(cluster_spear, 
                          ~melt(.x$p))
cluster_spear_padj <- map(cluster_spear, 
                          ~melt(.x$p.adj))
cluster_spear_melt <- pmap(list(cluster_spear_r, cluster_spear_p, cluster_spear_padj),
                           ~ reduce(list(..1,..2,..3), left_join, by = c("Var1","Var2")))
map(cluster_spear_melt, ~min(.x$value))
```



