---
title: "Clustering proteins"
output: html_notebook
---

```{r, echo=FALSE, include=FALSE}
library(readxl)
library(reshape2)
library(psych)
library(phyloseq)
library(dendextend)
library(gplots)
library(clusterProfiler)
library(CARlasso)
library(tidyverse)
library(ggprism)
theme_set(theme_prism(base_size = 14))
```

Import all the necessary matrices for analysis
```{r import}
files <- list.files(path = here::here("corr_mats/"))
ms_file_list <- str_subset(files, "ms")
#Import MS data
mq_mat <- read_csv(file = here::here(str_c("corr_mats/",ms_file_list))) %>% 
  column_to_rownames(var = "...1") %>% 
  as.matrix()
```
We will be doing correlations within just Ileum or Colon and within noIBD, active, or inactive IBD
```{r import-meta}
meta_full <- read_excel(here::here("Metadata.xlsx")) %>% 
  select("sample" = `#SampleID`, "group" = Group, everything()) %>% 
  mutate(sample = str_remove(sample, "Rep")) %>% 
  filter(sample %in% rownames(mq_mat)) %>% 
  mutate(across(matches("(Active)|(TI_resection)|(Cholestyramine)"), ~as.character(fct_recode(as_factor(.x), 
                                                Yes = "1", No = "0", Not_applicable = "NA"))))
#Prepare metadata tables
meta_group_list <- vector(mode = "list", length = 4) %>% 
  map(~ meta_full) %>%  
  map2(rep(c("Colon","Ileum"), each = 2), 
       ~filter(.x, Tissue_type == .y)) %>% 
  map2(rep(c("Control","IBD"),times = 2), 
       ~filter(.x, group == .y)) %>% 
  set_names(nm = str_c(rep(c("Colon","Ileum"), each = 2), rep(c("Control","IBD"),times = 2), sep = "_"))
meta_active_list <- vector(mode = "list", length = 6) %>% 
  map(~ meta_full) %>%  
  map2(rep(c("Colon","Ileum"), each = 3), 
       ~filter(.x, Tissue_type == .y)) %>% 
  map2(rep(c("Not_applicable","No","Yes"),times = 2), 
       ~filter(.x, Active_IBD == .y)) %>% 
  set_names(str_c(rep(c("Colon","Ileum"), each = 3),rep(c("Not_applicable","No","Yes"),times = 2), sep = "_"))
meta_list <- c(meta_group_list, meta_active_list)

#Keep list names for convenience
subgroups <- names(meta_list)
```
Further filter the proteomics data to just include DE proteins between active/inactive IBD vs control in each group
```{r select_top_prots}
#Import the results
da_res <- map(list.files(here::here("spear_subset_on_Foster"), pattern = "full_"),
              ~read_csv( file = here::here(str_c("spear_subset_on_Foster/",.)))) 
da_res <- set_names(da_res, nm = list.files(here::here("spear_subset_on_Foster"), pattern = "full_"))
#Reduce the results to one table, and filter afterwards
#Filter for significance and effect 
da_res_interest <- da_res %>% 
  map(filter, if_any(contains("adj"), ~.<0.01), if_any(contains("eff"), ~abs(.)>0.5)) %>% 
  reduce(full_join, by = c("gene.name","first.id")) %>% 
  select(gene.name,first.id, contains("padj"), -contains("grand"), contains("eff")) 


#Get the protein names to select
da_prots <- da_res_interest %>% 
  select(gene.name) %>% 
  separate(gene.name, into = letters [1:20], sep = " ") %>% 
  pivot_longer(letters [2:20], names_to = "letters", values_to = "secondary_names")
da_prots_final <- c(da_prots$a,da_prots$secondary_names) %>% 
  . [!is.na(.)] %>% 
  str_remove_all(";")

#Create the new mq_list
mq_top_list <- map(meta_list, ~ mq_mat [,colnames(mq_mat) %in% da_prots_final]) %>% 
  map2(meta_list, ~ .x [pull(.y,var = sample),]) 
```




Run correlations on the top proteins in each comparison - only look at IBD colon and ileum for now

```{r spearman}
mq_select_top <- mq_top_list [c(2,4)]
#Run the spearman correlations
psych_cor_list <- mq_select_top %>% 
  map(corr.test, method = "spearman")
#Extract the coef and p-value matrices
psych_cor_mats_list <- psych_cor_list %>% 
  map(function(x) map(c("r","p"), 
                        ~x [.x]))
#Melt the matrices
#Create the helper function to assign NA to lower triangle of the matrix
assign_lower_tri_na <- function(x) {
  x [lower.tri(x, diag = T)] <- NA
  return(x)
}
psych_dfs_list <- psych_cor_mats_list %>% 
  map_depth(1, flatten) %>%  
  map_depth(2, assign_lower_tri_na) %>%  
  map_depth(2, melt) %>% 
  map_depth(2, filter, !is.na(value)) %>% 
  map_depth(1, reduce, left_join, by = c("Var1","Var2")) %>% 
  map(rename, r = value.x, p = value.y)
```

Now that correlation is done, we can run the clustering analysis with Ward linkages. Roughly following (this)[https://bio723-class.github.io/Bio723-book/clustering-in-r.html#combining-heatmaps-and-dendrograms] for the clustering analysis. 

```{r clustering}
#Transform the correlation into Distances
  
psych_coef_list <- psych_cor_list %>% 
  map(~.x$r) 
  
psych_dist <- psych_coef_list %>% 
  map(~as.dist(1-.x))

#Use Ward linkages to cluster the correlation data

psych_tree <- psych_dist %>% 
  map(hclust, method = "ward.D2")

psych_dend <- psych_tree %>% 
  map(as.dendrogram)


map(psych_dend, nleaves)
map(psych_dend, nnodes)
#Try to cut the dendrogram into clusters - select arbitrarily based on the correlation heatmap
cluster_number <- set_names(c(8,7), nm = names(psych_dend))
psy_clust <- psych_dend %>% 
  map2(cluster_number, ~cutree(.x,k=.y))
#Check how many proteins in each cluster
map(psy_clust, table)

pwalk(list(psych_dend,names(psych_dend),cluster_number), ~plot(color_branches(..1, k=..3), leaflab = "none", xlab = ..2 ))
```

We can try visualizing the results in a heatmap of correlations

```{r k-medoid-cluster, fig.width=6, fig.height=6}
#Create clusters using k-medoids
psy_kmed <- psych_dist %>% 
  map2(cluster_number, ~cluster::pam(.x,k=.y))
map(psy_kmed, ~table(.x$cluster))
psy_kmed_clusters <- psy_kmed %>% 
  map(~.x$cluster)
```


```{r pheatmap, fig.width=9, fig.height=9}
#Set the heatmap color scheme
mycolors <- rcartocolor::carto_pal(name = "Geyser")
pal_length <- 255
color_sheme <- colorRampPalette(mycolors) (pal_length)
#Set the color scheme for the cluster annotation
cluster_colors <- cluster_number %>% 
  map(~list(Cluster = set_names(rcartocolor::carto_pal(name = "Vivid") [1:.x],
                                           nm = str_c("Cluster",1:.x, sep = "_"))))

#Set the breaks to be able to center the heatmap colors around 0
mybreaks_list <- psych_coef_list %>% 
  map(~c(seq(min(.x), 0, length.out=ceiling(pal_length/2) + 1), 
              seq(max(.x)/pal_length, max(.x), length.out=floor(pal_length/2))))

#Prepare clustering according to the Ward method
psy_ward_cluster_annot <- psy_clust %>% 
  map(enframe) %>% 
  map(column_to_rownames, var = "name") %>% 
  map(transmute,Cluster = str_c("Cluster",value, sep = "_")) 
#Prepare clustering according to k-medoid method and re-order the correlation matrix according to the k-medoid clustering
psy_kmed_cluster_annot <- psych_dist %>% 
  map2(cluster_number, ~cluster::pam(.x,k=.y,cluster.only = T)) %>% 
  map(enframe) %>% 
  map(column_to_rownames, var = "name") %>% 
  map(transmute,Cluster = str_c("Cluster",value, sep = "_")) 
psych_kmed_coef_list <- pmap(list(psych_coef_list,psych_dist,cluster_number),
                             ~ ..1 [order(cluster::pam(..2,k=..3, cluster.only = T)),
                                   order(cluster::pam(..2,k=..3, cluster.only = T))])

#Print out the heatmaps
#pdf(here::here("protein_cluster_correlation/DA_clustering_heatmaps.pdf"), width = 9, height = 9)
pwalk(list(psych_coef_list, names(psych_coef_list), mybreaks_list),
     ~pheatmap::pheatmap(..1, color = color_sheme,
                   cluster_rows = F, cluster_cols = F, breaks = ..3,
                   treeheight_row = 0, treeheight_col = 0,
                   main = str_c("No clustering", ..2, sep = "\n"), fontsize = 5))
pwalk(list(psych_coef_list, names(psych_coef_list), mybreaks_list, psych_tree, psy_ward_cluster_annot, cluster_colors),
      ~pheatmap::pheatmap(..1, color = color_sheme, breaks = ..3,
                   cluster_rows = ..4, cluster_cols = ..4,
                   annotation_col = ..5, treeheight_row = 0, treeheight_col = 0,
                   main = str_c("Ward",..2,sep = "\n"), fontsize = 5,
                   annotation_colors = ..6))

pwalk(list(psych_kmed_coef_list, names(psych_kmed_coef_list), mybreaks_list, psy_kmed_cluster_annot, cluster_colors),
      ~pheatmap::pheatmap(..1, color = color_sheme,
                   cluster_rows = F, cluster_cols = F, breaks = ..3,
                   annotation_col = ..4, treeheight_row = 0, treeheight_col = 0,
                   main = str_c("K-Medoid", ..2, sep = "\n"), fontsize = 5,
                   annotation_colors = ..5))
#dev.off()
```

Perform functional analysis of the clusters
```{r cluster-profiler}
h <- read.gmt(here::here("h.all.v7.5.1.symbols.gmt"))%>% 
  mutate(term = str_remove(term,"HALLMARK_"))
cp <- read.gmt(here::here("c2.cp.v7.5.1.symbols.gmt")) %>% 
  filter(str_detect(term, "REACT")) %>% 
  mutate(term = str_remove(term,"REACTOME_"))
universe <- colnames(mq_mat)
#Get lists of Ward clusters from clustering data
psy_ward_cluster_dfs <- psy_clust %>% 
  map(melt) %>% 
  map(rownames_to_column, var = "protein") 
psy_ward_clusters <- psy_ward_cluster_dfs %>% 
  map(nest, data = "protein") %>% 
  map(~ .x [[2]]) %>% 
  map_depth(2, flatten_chr) %>%  
  map_depth(1, ~set_names(.x, nm = str_c("cluster", 1:length(.x), sep = "_")))
psy_ward_clusters_csv <- psy_ward_clusters %>% 
  map(~list2DF(.x)) %>% 
  map(~mutate(.x,across(everything(), .fns = ~if_else(duplicated(.x), "NA", .x)))) %>% 
  map(mutate, across(everything(), str_remove, "^NA$"))
#walk2(psy_ward_clusters_csv, names(psy_ward_clusters_csv),
 #     ~write_csv(.x,here::here(str_c("protein_cluster_correlation/",.y,"_protein_clusters.csv"))))

#Perform ORA analysis using either Hallmark or Reactome pathways

h_list <- psy_ward_clusters %>% 
  map_depth(2, 
            ~enricher(.x, TERM2GENE = h, 
                      pvalueCutoff = 0.05, pAdjustMethod = "BH", universe = universe,
                      qvalueCutoff = 0.05))

cp_list <- psy_ward_clusters %>% 
  map_depth(2, 
            ~enricher(.x, TERM2GENE = cp, 
                      pvalueCutoff = 0.05, pAdjustMethod = "BH", universe = universe,
                      qvalueCutoff = 0.05))

#Filter to only keep succesful clusters

suc_h_list <- h_list %>% 
  map(~compact(.x)) %>% 
  map(~keep(.x,~nrow(.x) > 0))
flat_suc_h_list <- suc_h_list %>% 
  map2(names(suc_h_list), ~set_names(.x,nm=str_c(.y, names(.x), sep = "__"))) %>% 
  flatten() 
suc_cp_list <- cp_list %>% 
  map(~keep(.x,~nrow(.x) > 0))
flat_suc_cp_list <- suc_cp_list %>% 
  map2(names(suc_cp_list), ~set_names(.x,nm=str_c(.y, names(.x), sep = "__"))) %>% 
  flatten() 
```

Visualize the results of functional profiling

```{r cluster-profiler-vis}
#Visualize functional results
h_dot_list <- flat_suc_h_list %>% 
  map2(names(flat_suc_h_list),~dotplot(.x, showCategory = 30) +
        scale_color_gradientn(colours = mycolors [7:4]) +
        labs(caption = .y, subtitle = "Hallmark"))
cp_dot_list <- flat_suc_cp_list %>% 
  map2(names(flat_suc_cp_list),~dotplot(.x, showCategory = 20) +
        scale_color_gradientn(colours = mycolors [7:4]) +
        labs(caption = .y, subtitle = "Reactome") +
         theme(axis.text.y = element_text(size = 5)))
h_heat_list <- flat_suc_h_list %>% 
  map2(names(flat_suc_h_list),~heatplot(.x, showCategory = 30) +
        labs(caption = .y, subtitle = "Hallmark"))

cp_heat_list <- flat_suc_cp_list %>% 
  map2(names(flat_suc_cp_list),~heatplot(.x, showCategory = 20) +
        labs(caption = .y, subtitle = "Reactome") +
         theme(axis.text.y = element_text(size = 5)))

#Combine the results into lists for grid plotting
h_dot_heat_list <- map2(h_dot_list, h_heat_list, ~set_names(list(.x,.y), nm = c("dot","heat")))
cp_dot_heat_list <- map2(cp_dot_list, cp_heat_list, ~set_names(list(.x,.y), nm = c("dot","heat")))
```

Plot the results on a grid
```{r cluster-profiler-grid, fig.width=6, fig.height=8} 
#pdf(here::here("protein_cluster_correlation/Hallmark_enrichment.pdf"), width = 6, height = 8)
map(h_dot_heat_list,
  ~cowplot::plot_grid(plotlist = .x, nrow = 2))
#dev.off()
#pdf(here::here("protein_cluster_correlation/Reactome_enrichment.pdf"), width = 6, height = 12)
map(cp_dot_heat_list,
  ~cowplot::plot_grid(plotlist = .x, nrow = 2))
#dev.off()
```
Next step is to average expression of the proteins within a cluster (only keep clusters that had some functionality identified) and correlate that with bacterial expression

```{r average_cluster_expression}
mq_cluster_expression <- mq_select_top %>% 
  map(melt) %>% 
  map(rename, protein = Var2, sample = Var1, abundance = value) %>% 
  map2(psy_ward_cluster_dfs, left_join, by = "protein") %>% 
  map(mutate, value = str_c("cluster",value, sep = "_")) %>% 
  map(group_by, value, sample) %>% 
  map(summarize, cluster_mean = mean(abundance)) 

mq_relevant_cluster_expr <- pmap(list(mq_cluster_expression, suc_h_list, suc_cp_list),
                                 ~ filter(..1, value %in% union(names(..2), names(..3))))

mq_cluster_mat <- mq_relevant_cluster_expr %>% 
  map(pivot_wider, names_from = value, values_from = cluster_mean) %>% 
  map(column_to_rownames, var = "sample") %>% 
  map(as.matrix)
```

Next get bacterial families
Import all the necessary matrices for analysis
```{r import-otu}
files <- list.files(path = here::here("corr_mats/"))
otu_file_list <- str_subset(files,"clr(.)+otu")
tax_file_list <- str_subset(files,"tax")
#Import otu counts
otu_list <- map(otu_file_list, 
                    ~read_csv(file = here::here(str_c("corr_mats/",.))))
otu_mat_list <- otu_list %>% 
  map(column_to_rownames, var = "...1") %>% 
  map(as.matrix)
#Only keep family level otus - tried that - no success - most signficant correlation was ileum-ileum p=0.1
#Do this at the genus level -
otu_genus <- otu_mat_list [[2]]
#Prepare the OTU matrices for each comparison
otu_fam_list <- mq_cluster_mat %>% 
  map(~ otu_genus [rownames(.x),])
```


```{r select-top-otu}
biom <-  import_biom("table-with-taxonomy.biom", parseFunction = parse_taxonomy_default)
tree <- read_tree_greengenes("tree.nwk")
metadata <- import_qiime_sample_data("metadata.txt")
physeq <- merge_phyloseq(biom, metadata, tree)
colnames(tax_table(physeq)) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
#Remove irrelevant samples
physeq <- subset_samples(physeq, !(sample_names(physeq) %in% c("C34C","I01I","I60I","I12I", "I44I")))
#Remove all mocks, and negative controls
physeq <- subset_samples(physeq, str_detect(X.SampleID, "(Mock)|(Neg)|(TC7)|(H2O)", negate = T))
sample_names(physeq) <- str_remove(sample_names(physeq),"Rep")

phy_fam <- tax_glom(physeq, taxrank = "Genus")
phy_fam_rel <- transform_sample_counts(phy_fam, function(x) x/sum(x)) 

otu_fam_rel <- mq_select_top %>% 
  map(~otu_table(phy_fam_rel)) %>% 
  map2(mq_select_top, ~.x [,rownames(.y)])


#In each case select a top n number of families (mean abundance)
top_cut <- 40
otu_top_fam <- otu_fam_rel %>%
  map(melt) %>% 
  map(group_by, Var1) %>%   
  map(summarise, mean_abund = mean(value)) %>% 
  map(~arrange(.x,desc(mean_abund))) %>% 
  map(slice_head, n=top_cut)

otu_fam_list <- mq_cluster_mat %>% 
  map(~ otu_genus [rownames(.x),])

otu_fam_list <- otu_fam_list %>% 
  map2(otu_top_fam, ~.x [,colnames(.x) %in% pull(.y,var = "Var1")])
```

Run Spearman correlations on top genera and proteins
```{r cluster-otu-spear} 
#Set up the matrices for correlations
cluster_fam_list_names <- cross2(names(mq_cluster_mat), names(otu_fam_list)) %>% 
  map(reduce,str_c, sep = "__protein:bacterial__")
cluster_fam_list <- cross2(mq_cluster_mat, otu_fam_list) %>% 
  set_names(cluster_fam_list_names)
#Run the correlations and melt the dataframe
cluster_spear <- map(cluster_fam_list,
                      ~corr.test(x=.x [[1]],y=.x [[2]], method = "spearman", adjust = "BH"))
cluster_spear_r <- map(cluster_spear, 
                          ~melt(.x$r))
cluster_spear_p <- map(cluster_spear, 
                          ~melt(.x$p))
cluster_spear_padj <- map(cluster_spear, 
                          ~melt(.x$p.adj))
cluster_spear_melt <- pmap(list(cluster_spear_r, cluster_spear_p, cluster_spear_padj),
                           ~ reduce(list(..1,..2,..3), left_join, by = c("Var1","Var2")))
#Check the lowest adjusted p value
map_chr(cluster_spear_melt, ~min(.x$value))
#Get the tax table out for nicer names
tax <- as.data.frame(tax_table(phy_fam)) %>% 
  rownames_to_column(var = "Var2")
#Prepare the correlation table for plotting
cluster_spear_plot <- cluster_spear_melt %>% 
  map(filter, value < 0.15) %>% 
  keep(~nrow(.x)>0) %>% 
  map(left_join, tax, by = "Var2")

genus_cluster_corplot <- map2(cluster_spear_plot, names(cluster_spear_plot),
                              ~ggplot(.x, aes(x=Var1, y = Genus)) +
                                geom_point(aes(color = value.x, size = round(value,3))) +
                                scale_color_gradient2(low = mycolors [1], mid = mycolors [4], high = mycolors[7],
                                                      midpoint = 0, name = "Spearman's R") +
                                labs(x = "Protein Cluster", y=element_blank(), title = .y, size = "padj") + 
                                theme_classic(base_size = 12))
```


```{r cluster-otu-spear-plot-grid, fig.width=6, fig.height=6} 
cowplot::plot_grid(plotlist = genus_cluster_corplot, nrow = 2)
#ggsave(here::here("protein_cluster_correlation/genus_protein_cluster_cor.png"), width = 6, height = 6, dpi = 1080)
```

Try and make clusters of bacteria and correlate those clusters with protein clusters
```{r genera-clusters}
#Convert the feature names into genus names
genera_list <- otu_fam_list %>% 
  map(melt) %>% 
  map(left_join,tax, by = "Var2") %>% 
  map(mutate, true_genus = if_else(str_detect(Genus, "uncultured"), Family, Genus)) %>% 
  map(select,Var1, true_genus, value) %>% 
  map(pivot_wider, names_from = true_genus, values_from = value) %>% 
  map(column_to_rownames, var = "Var1") %>% 
  map(as.matrix)
#Run the spearman correlations
psych_cor_genus_list <- genera_list %>% 
  map(corr.test, method = "spearman")
#Extract the coef and p-value matrices
psych_cor_mats_genus_list <- psych_cor_genus_list %>% 
  map(function(x) map(c("r","p"), 
                        ~x [.x]))
#Melt the matrices

psych_dfs_genus_list <- psych_cor_mats_genus_list %>% 
  map_depth(1, flatten) %>%  
  map_depth(2, assign_lower_tri_na) %>%  
  map_depth(2, melt) %>% 
  map_depth(2, filter, !is.na(value)) %>% 
  map_depth(1, reduce, left_join, by = c("Var1","Var2")) %>% 
  map(rename, r = value.x, p = value.y)
#Transform the correlation into Distances
  
psych_coef_genus_list <- psych_cor_genus_list %>% 
  map(~.x$r) 
  
psych_dist_genus <- psych_coef_genus_list %>% 
  map(~as.dist(1-.x))

#Use Ward linkages to cluster the correlation data

psych_tree_genus <- psych_dist_genus %>% 
  map(hclust, method = "ward.D2")

psych_dend_genus <- psych_tree_genus %>% 
  map(as.dendrogram)


map(psych_dend_genus, nleaves)
map(psych_dend_genus, nnodes)
#Try to cut the dendrogram into clusters - select arbitrarily based on the correlation heatmap
cluster_number_genus <- set_names(c(4,4), nm = names(psych_dend_genus))
psy_clust_genus <- psych_dend_genus %>% 
  map2(cluster_number_genus, ~cutree(.x,k=.y))
#Check how many proteins in each cluster
map(psy_clust_genus, table)

pwalk(list(psych_dend_genus,names(psych_dend_genus),cluster_number_genus), ~plot(color_branches(..1, k=..3), leaflab = "none", xlab = ..2 ))
```
Look at bacterial correlational heatmaps
```{r pheatmap-genera, fig.width=8, fig.height=8}
#Set the breaks to be able to center the heatmap colors around 0
mybreaks_list_genus <- psych_coef_genus_list %>% 
  map(~c(seq(min(.x), 0, length.out=ceiling(pal_length/2) + 1), 
              seq(max(.x)/pal_length, max(.x), length.out=floor(pal_length/2))))
#Set the color scheme for the cluster annotation
cluster_colors_genus <- cluster_number_genus %>% 
  map(~list(Cluster = set_names(rcartocolor::carto_pal(name = "Vivid") [1:.x],
                                           nm = str_c("Cluster",1:.x, sep = "_"))))

#Prepare clustering according to the Ward method
psy_ward_cluster_annot_genus <- psy_clust_genus %>% 
  map(enframe) %>% 
  map(column_to_rownames, var = "name") %>% 
  map(transmute,Cluster = str_c("Cluster",value, sep = "_")) 

#Print out the heatmaps
#pdf(here::here("protein_cluster_correlation/top_genera_clustering_heatmaps.pdf"), width = 8, height = 8)
pwalk(list(psych_coef_genus_list, names(psych_coef_genus_list), mybreaks_list_genus),
     ~pheatmap::pheatmap(..1, color = color_sheme,
                   cluster_rows = F, cluster_cols = F, breaks = ..3,
                   treeheight_row = 0, treeheight_col = 0,
                   main = str_c("No clustering", ..2, sep = "\n"), fontsize = 7))
pwalk(list(psych_coef_genus_list, names(psych_coef_genus_list), mybreaks_list_genus, 
           psych_tree_genus, psy_ward_cluster_annot_genus, cluster_colors_genus),
      ~pheatmap::pheatmap(..1, color = color_sheme, breaks = ..3,
                          cluster_rows = ..4, cluster_cols = ..4,
                          annotation_col = ..5, treeheight_row = 0, treeheight_col = 0,
                          main = str_c("Ward",..2,sep = "\n"), fontsize = 7,
                          annotation_colors = ..6))

#dev.off()
```
Bacteria do not cluster that nicely, so keep them in separate genera. Finally, try to analyze the data with CARLASSO - families with protein clusters in colon or IBD
```{r carlasso-otu-prep}
#Prepare the bacterial family dataframes
phy_fam_carlasso <- tax_glom(physeq, taxrank = "Family")
tax_fam <- as.data.frame(tax_table(phy_fam_carlasso)) %>% 
  rownames_to_column(var = "feature")
otu_fam_carlasso <- as.data.frame(otu_table(phy_fam_carlasso)) %>% 
  rownames_to_column(var = "feature") 
otu_fam_carlasso_rel <- otu_fam_carlasso %>% 
  mutate(across(where(is.numeric), function(x) x/sum(x)))
#Select number of families to keep
tokeep <- 10
#Create two matrices - one for colon and one for ileum
otu_fam_carlasso_rel_list <- set_names(c("Colon", "Ileum"),c("Colon", "Ileum")) %>% 
  map(~otu_fam_carlasso_rel [,c("feature", pull(filter(meta_full, Tissue_type == .x), var = "sample"))])
otu_fam_carlasso_rel_melt_list <- otu_fam_carlasso_rel_list %>% 
  map(melt) %>% 
  map(group_by, feature) %>% 
  map(summarise, mean_abund = mean(value)) %>% 
  map(arrange, desc(mean_abund))

#Select families to keep and families to aggregate
fam_to_keep <- otu_fam_carlasso_rel_melt_list %>% 
  map(~pull(slice_head(.x, n=tokeep), feature))
fam_others <- otu_fam_carlasso_rel_melt_list %>% 
  map(~pull(slice_tail(.x, n=-tokeep), feature))

#Prepare dataframes of just top bugs and the others aggregates
otu_fam_carlasso_named <- otu_fam_carlasso_rel_list %>% 
  map(~otu_fam_carlasso [,colnames(.x)]) %>%
  map2(fam_to_keep, ~ filter(.x, feature %in% .y))
otu_fam_carlasso_others <- otu_fam_carlasso_rel_list %>% 
  map(~otu_fam_carlasso [,colnames(.x)]) %>%
  map2(fam_others, ~ filter(.x, feature %in% .y)) %>% 
  map(summarise,across(where(is.numeric), sum)) %>%  
  map(mutate, feature = "all_others")
otu_fam_carlasso_list <- otu_fam_carlasso_named %>% 
  map2(otu_fam_carlasso_others, bind_rows)
#Replace feature ids with meaningful name
otu_fam_carlasso_better_list <- otu_fam_carlasso_list %>% 
  map(left_join, tax_fam, by = "feature") %>% 
  map(mutate, final_name = if_else(is.na(Family),feature, Family))
#Add metadata to the dataframes
otu_fam_meta_carlasso <- otu_fam_carlasso_better_list %>% 
  map(select, -feature, -Kingdom, -Phylum, -Order, -Class, -Family, -Genus, -Species) %>% 
  map(melt) %>% 
  map(left_join, select(meta_full, sample, group, TI_resection, Active_TI, Active_RC, Cholestyramine, Active_IBD),
      by = c("variable" = "sample"))
```
Now get the proteomics data and prepare that for analysis
```{r carlasso-mq-prep}
mq_raw <- read_tsv(here::here("../../Proteomics/normalyzer/Global_group/submitted_rawdata.txt"))
#Keep only samples for which other data exist
mq_raw_organized <- mq_raw %>% 
  select(X, all_of(rownames(mq_mat)))
#Separate the data into colon and ileum
mq_raw_list <- otu_fam_meta_carlasso %>% 
  map(~mq_raw_organized [, c("X", unique(.x$variable))])
#Melt and convert into cluster abundances
mq_raw_clust_list <- mq_raw_list %>% 
  map(melt) %>% 
  map(rename, protein = X, abundance = value, sample = variable) %>% 
  map2(psy_ward_cluster_dfs, left_join, by = "protein") %>% 
  map(filter, !is.na(value)) %>% 
  map(mutate, value = str_c("cluster",value, sep = "_")) %>% 
  map(group_by, value, sample) %>% 
  map(summarize, cluster_mean = mean(abundance, na.rm = T)) 
#Only keep successful clusters with some functional annotation
mq_raw_suc_clust_list <- pmap(list(mq_raw_clust_list, suc_h_list, suc_cp_list),
                                 ~ filter(..1, value %in% union(names(..2), names(..3))))
#Add metadata to the MS data frame and proceed to carlasso analysis
mq_suc_clust_meta_carlasso <- mq_raw_suc_clust_list %>% 
  map(left_join, select(meta_full, sample, group, TI_resection, Active_TI, Active_RC, Cholestyramine, Active_IBD),
      by = "sample") %>% 
  map(rename, final_name = value, variable = sample, value = cluster_mean)
```
Now ready for Carlasso analysis
```{r carlasso-setup}
carlasso_df_long_list <- otu_fam_meta_carlasso %>% 
  map2(mq_suc_clust_meta_carlasso, bind_rows)
#Make the final carlasso dfs
carlasso_final_df_list <- carlasso_df_long_list %>% 
  map(pivot_wider, names_from = final_name, values_from = value) %>% 
  map(mutate, 
      TI_resection = fct_relevel(TI_resection, "Yes","Not_applicable", "No"),
      Active_IBD = fct_relevel(Active_IBD, "Yes", "Not_applicable", "No"),
      Active_RC = fct_relevel(Active_RC,"Yes", "Not_applicable", "No"),
      Active_TI = fct_relevel(Active_TI,"Yes", "Not_applicable", "No"))
#Prepare the carlasso formula
carlasso_lhs_formula_list <- carlasso_final_df_list %>% 
  map(~colnames(.x)) %>% 
  map(~ .x [str_detect(.x,"f__") | str_detect(.x,"cluster")]) %>% 
  map(~ c(.x, "all_others")) %>% 
  map(str_c, collapse = " + ")
carlasso_rhs_formula_list <- carlasso_final_df_list %>% 
  map(~colnames(.x)) %>% 
  map(~ .x [str_detect(.x,"f__", negate = T) & 
              str_detect(.x,"cluster", negate = T) &
              str_detect(.x,"all_others", negate = T) &
              str_detect(.x,"variable", negate = T)]) %>%
  map(str_c, collapse = " + ")
carlasso_formula_list <- carlasso_lhs_formula_list %>% 
  map2(carlasso_rhs_formula_list, str_c, sep = " ~ ")
#Finally, run carlasso
classo_res_list <- carlasso_final_df_list %>% 
  map2(carlasso_formula_list, 
       ~CARlasso(as.formula(.y),
                    data = .x,link = "logit", 
                    adaptive = TRUE, n_iter = 2000, 
                    n_burn_in = 1000, thin_by = 2))
classo_res_hs_list <- classo_res_list %>% 
  map(horseshoe)
```
Now can plot the results
```{r carlasso-plot, fig.width=11, fig.height=11}
#pdf(here::here("protein_cluster_correlation/carlasso_protein_cluster_top_10_families.pdf"), width = 12, height = 12)
classo_res_hs_list %>% 
  map2(names(classo_res_hs_list),~plot(.x) +
         ggraph::scale_edge_color_manual(values = mycolors [c(1,7)]) +
         ggtitle(.y))
#dev.off()
```



