---
title: "Mixomics-DIABLO-starter"
output: html_notebook
author: Zakhar Krekhno
---

```{r, echo=FALSE, include=FALSE}
library(phyloseq)
library(mixOmics)
library(readxl)
library(reshape2)
library(tidyverse)
theme_set(theme_classic(base_size = 14))
```

Import all the necessary data

```{r import}
files <- list.files(path = here::here("corr_mats/"))
otu_file_list <- str_subset(files,"clr(.)+otu")
tax_file_list <- str_subset(files,"tax")
ms_file_list <- str_subset(files, "ms")
#Import otu counts
otu_list <- map(otu_file_list, 
                    ~read_csv(file = here::here(str_c("corr_mats/",.))))
otu_mat_list <- otu_list %>% 
  map(column_to_rownames, var = "...1") %>% 
  map(as.matrix)
#Import MS data
mq_mat <- read_csv(file = here::here(str_c("corr_mats/",ms_file_list))) %>% 
  column_to_rownames(var = "...1") %>% 
  as.matrix()
```
```{r import-meta}
meta_full <- read_excel(here::here("Metadata.xlsx")) %>% 
  select("sample" = `#SampleID`, "group" = Group, everything()) %>% 
  mutate(sample = str_remove(sample, "Rep")) %>% 
  filter(sample %in% rownames(mq_mat)) %>% 
  mutate(across(matches("(Active)|(TI_resection)|(Cholestyramine)"), ~as.character(fct_recode(as_factor(.x), 
                                                Yes = "1", No = "0", Not_applicable = "NA"))))
#Prepare metadata tables
meta_list_starter <- map(vector(mode = "list", length = 6), ~ meta_full)
meta_list <-meta_list_starter %>% 
  map2(rep(c("Colon","Ileum"), each = 3), 
       ~filter(.x, Tissue_type == .y)) %>% 
  map2(rep(c("Not_applicable","No","Yes"),times = 2), 
       ~filter(.x, Active_IBD == .y))
#Name the list for convenience
subgroups <- str_c(rep(c("Colon","Ileum"), each = 3),rep(c("Not_applicable","No","Yes"),times = 2), sep = "_")
names(meta_list) <- subgroups
```
Make appropriate matrices for each group
```{r subset_mq}
mq_list <- map(vector(mode = "list", length = 6*3), ~ mq_mat) %>% 
  map2(rep(meta_list, each=3), ~ .x [rownames(.x) %in% pull(.y,var = sample),])
otu_filt_list <- rep(otu_mat_list, times = 6) %>% 
  map2(rep(meta_list, each=3), ~ .x [rownames(.x) %in% pull(.y,var = sample),])
```
Further filter the proteomics data to just include DE proteins between active/inactive IBD vs control in each group
```{r select_top_prots}
prot_da_res <- read_excel(here::here("All comparisons vsn-norm limma cut-off 0.5 robust.xlsx"))
#Select proteins that were significantly DA in at least one comparison
prot_da_interest <- prot_da_res %>% 
  select(Gene,contains("adj")) %>% 
  filter(if_any(contains("adj"), ~.<0.01))
#Create the new mq_list
mq_top_list <- map(vector(mode = "list", length = 6*3), ~ mq_mat [,colnames(mq_mat) %in% prot_da_interest$Gene]) %>% 
  map2(rep(meta_list, each=3), ~ .x [rownames(.x) %in% pull(.y,var = sample),])
```

Attempt DIABLO analysis

```{r prep-diablo}
#Create a factor for separation based on ActiveIBD status
Y <- filter(meta_full, Tissue_type == "Ileum") %>% 
  pull(Active_IBD) %>% 
  as_factor()
otu <- otu_mat_list [[1]] %>% 
  . [pull(filter(meta_full, Tissue_type == "Ileum"),sample),]
mq <- mq_mat [,colnames(mq_mat) %in% prot_da_interest$Gene] %>% 
  . [pull(filter(meta_full, Tissue_type == "Ileum"),sample),] 
#Ensure rows are the same order
all(rownames(otu) == rownames(mq))
X <- list(protein = mq, microbiome = otu)

#Run starter DIABLO
res.dblo <- block.splsda(X, Y)
plotIndiv(res.dblo, ind.names = TRUE, 
          legend=TRUE, cex=c(1,2,3))
plotVar(res.dblo, var.names = c(TRUE, FALSE),
        legend=TRUE)
plotDiablo(res.dblo, ncomp = 1)
#circosPlot(res.dblo, cutoff=0.7)
cimDiablo(res.dblo, margin=c(8,20))
cimDiablo(res.dblo, color.blocks = c('darkorchid', 'lightgreen'), comp = 1, margin=c(8,20), legend.position = "right")
plotLoadings(res.dblo, comp = 1, contrib = "max")
```
```{r}
set.seed(123) # for reproducibility in this vignette
MyPerf.diablo <- perf(res.dblo, validation = 'Mfold', folds = 5, 
                   nrepeat = 10, 
                   dist = 'centroids.dist')

MyPerf.diablo  # lists the different outputs

# Performance with Majority vote
MyPerf.diablo$MajorityVote.error.rate
Myauc.diablo <- auroc(res.dblo, roc.block = "microbiome", roc.comp = 2)
```



