---
title: "Spearman correlations"
output: html_notebook
author: Zakhar Krekhno
---

```{r, echo=FALSE, include=FALSE}
library(phyloseq)
library(EnhancedVolcano)
library(readxl)
library(reshape2)
library(psych)
library(tidyverse)
theme_set(theme_classic(base_size = 14))
calc_relative_abund <- function(x) x/sum(x)
```

Import proteomics data
```{r import-prot}
mq <- read_excel(here::here("Rav_198results.xlsx"), sheet = 2)
#Log transform the MS data
mq_log <- mq %>% 
  mutate(across(-Genes,~ as.numeric(str_replace_na(.,replacement = 0)))) %>% 
  mutate(across(-Genes,~log2(.+1)))
codes <- read_excel(here::here("Metadata.xlsx"))
meta_full <- codes %>% 
  select("sample" = `#SampleID`, "group" = Group, everything()) %>% 
  filter(sample %in% colnames(mq)) %>% 
  mutate(across(matches("(Active)|(TI_resection)|(Cholestyramine)"), ~as.character(fct_recode(as_factor(.x), 
                                                Yes = "1", No = "0", Not_applicable = "NA"))))

```

Import 16S data
Filter the samples a bit:
Remove all irrelevant samples. Sample I60IRep seems to actually be sample I61I, so rename it as such. Remove REP from sample names

```{r import-16S}
biom <-  import_biom("table-with-taxonomy.biom", parseFunction = parse_taxonomy_default)
tree <- read_tree_greengenes("tree.nwk")
metadata <- import_qiime_sample_data("metadata.txt")
physeq <- merge_phyloseq(biom, metadata, tree)
colnames(tax_table(physeq)) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
#Remove irrelevant samples
physeq <- subset_samples(physeq, !(sample_names(physeq) %in% c("C34C","I01I","I60I","I12I", "I44I")))
#Remove all mocks, and negative controls
physeq <- subset_samples(physeq, str_detect(X.SampleID, "(Mock)|(Neg)|(TC7)|(H2O)", negate = T))
sample_names(physeq) <- str_remove(sample_names(physeq),"Rep")
#Choose only samples for which both 16S and proteomic data exist
samples <- intersect(colnames(mq),sample_names(physeq))
physeq_intersect <- subset_samples(physeq, sample_names(physeq) %in% samples)
mq_intersect <- mq_log %>% 
  column_to_rownames(var = "Genes") %>% 
  select(all_of(samples))
#Check if samples have both proteomic and 16S data
all(colnames(mq_intersect) %in% sample_names(physeq_intersect))
all(sample_names(physeq_intersect) %in% colnames(mq_intersect))
```
Next transpose the matrices to prepare them for Spearman correlations

```{r create-spearman-matrix}
#Prepare the MS data matrix for spearman correlations
mq_mat <- mq_intersect %>% 
  as.matrix() %>% 
  t()
#write.csv(mq_mat, here::here("corr_mats/ms_matrix.csv"))
#Prepare the 16S tables
phy_genus <- tax_glom(physeq_intersect,taxrank = "Genus")
phy_family <- tax_glom(physeq_intersect,taxrank = "Family")
#Calculate relative abundances for each phyloseq object
phy_rel <- map(list(physeq_intersect, phy_genus, phy_family),
               transform_sample_counts,calc_relative_abund)
#Extract otu and tax tables and save everything to csv files
tax_list <- map(phy_rel, tax_table)
#walk2(tax_list, c("Species_tax","Genus_tax", "Family_tax"),
 #     ~write.csv(.x, here::here(str_c("corr_mats/",.y,".csv", sep = ""))))

otu_rel <- map(phy_rel,otu_table)
otu_mat <- otu_rel %>% 
  map(as.matrix) %>% 
  map(t)
#walk2(otu_mat, c("Species_otu","Genus_otu", "Family_otu"),
 #     ~write.csv(.x, here::here(str_c("corr_mats/",.y,".csv", sep = ""))))
```




