---
title: "Spearman correlations prep"
output: html_notebook
author: Zakhar Krekhno
---

```{r, echo=FALSE, include=FALSE}
library(phyloseq)
library(readxl)
library(tidyverse)
calc_relative_abund <- function(x) x/sum(x)
```

Import proteomics data
```{r import-prot}
mq <- read_excel(here::here("Rav_198results.xlsx"), sheet = 2)
#Log transform the MS data
mq_log <- mq %>% 
  mutate(across(-Genes,~ as.numeric(str_replace_na(.,replacement = 0)))) %>% 
  mutate(across(-Genes,~log2(.+1)))
codes <- read_excel(here::here("Metadata.xlsx"))
meta_full <- codes %>% 
  select("sample" = `#SampleID`, "group" = Group, everything()) %>% 
  filter(sample %in% colnames(mq)) %>% 
  mutate(across(matches("(Active)|(TI_resection)|(Cholestyramine)"), ~as.character(fct_recode(as_factor(.x), 
                                                Yes = "1", No = "0", Not_applicable = "NA"))))

```

Import 16S data
Filter the samples a bit:
Remove all irrelevant samples. Sample I60IRep seems to actually be sample I61I, so rename it as such. Remove REP from sample names

```{r import-16S}
biom <-  import_biom("table-with-taxonomy.biom", parseFunction = parse_taxonomy_default)
tree <- read_tree_greengenes("tree.nwk")
metadata <- import_qiime_sample_data("metadata.txt")
physeq <- merge_phyloseq(biom, metadata, tree)
colnames(tax_table(physeq)) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
#Remove irrelevant samples
physeq <- subset_samples(physeq, !(sample_names(physeq) %in% c("C34C","I01I","I60I","I12I", "I44I")))
#Remove all mocks, and negative controls
physeq <- subset_samples(physeq, str_detect(X.SampleID, "(Mock)|(Neg)|(TC7)|(H2O)", negate = T))
sample_names(physeq) <- str_remove(sample_names(physeq),"Rep")
#Choose only samples for which both 16S and proteomic data exist
samples <- intersect(colnames(mq),sample_names(physeq))
physeq_intersect <- subset_samples(physeq, sample_names(physeq) %in% samples)
mq_intersect <- mq_log %>% 
  column_to_rownames(var = "Genes") %>% 
  select(all_of(samples))
#Check if samples have both proteomic and 16S data
all(colnames(mq_intersect) %in% sample_names(physeq_intersect))
all(sample_names(physeq_intersect) %in% colnames(mq_intersect))
```
Next transpose the matrices to prepare them for Spearman correlations

```{r create-spearman-matrix-ms}
#Prepare the MS data matrix for spearman correlations
#Only keep proteins that were detected in at least 30% of the samples above an expression cut-off of 6 (log-value)
#Remove low prevalence proteins
mq_prevalence_filter <- mq_intersect %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate(across(everything(), ~.>0))
mq_prev_sum <- mq_prevalence_filter %>% 
  summarise(across(everything(), sum)) %>% 
  mutate(across(everything(),~./ncol(mq_intersect)))
prevalent_proteins <- mq_prev_sum %>% 
  select(where(~.x >= 0.3)) %>% 
  colnames()
#Reduce the initial matrix to just prevalent proteins
mq_intersect_prev <- mq_intersect [rownames(mq_intersect) %in% prevalent_proteins,]
#Choose cut-off of 6
mq_prev_abund <- mq_intersect_prev [rowMeans(mq_intersect_prev) >6,]
mq_mat <- mq_prev_abund %>% 
  as.matrix() %>% 
  t()
#pdf(here::here("spear_results/MS_histograms.pdf"),width = 8, height = 6)
pwalk(list(list(mq_intersect,mq_intersect_prev,mq_prev_abund),
           c("Unfiltered","Prevalence-filtered", "Prevalence and abundance filtered"),
           c("","Prevalent in at least 30% of samples", "Mean abundance of at least 6 (log value)")),
      ~hist(rowMeans(..1),main = ..2, sub = ..3, xlab = "Log expression value"))
#dev.off()
#write.csv(mq_mat, here::here("corr_mats/ms_matrix.csv"))
```


```{r create-spearman-matrix-16s}
#Prepare the 16S tables
phy_genus <- tax_glom(physeq_intersect,taxrank = "Genus")
phy_family <- tax_glom(physeq_intersect,taxrank = "Family")
#Calculate relative abundances for each phyloseq object
phy_rel <- map(list(physeq_intersect, phy_genus, phy_family),
               transform_sample_counts,calc_relative_abund)
#Extract otu and tax tables and save everything to csv files
tax_list <- map(phy_rel, tax_table)
#walk2(tax_list, c("Species_tax","Genus_tax", "Family_tax"),
 #     ~write.csv(.x, here::here(str_c("corr_mats/",.y,".csv", sep = ""))))

otu_rel <- map(phy_rel,otu_table)
#Do similar filtering on otu tables as proteomics data - remove low prevalence and abundance microbes
otu_prevalent <- otu_rel %>% 
  map(t) %>% 
  map(as.data.frame) %>% 
  map(~mutate(.,across(everything(), ~.>0))) %>% 
  map(~summarise(.,across(everything(), sum))) %>% 
  map(~mutate(.,across(everything(),~./ncol(mq_intersect))))
#Keep bugs present in at least 5% of the patients
prevalent_otus_list <- otu_prevalent %>% 
  map(~select(.,where(~.x >= 0.05))) %>% 
  map(colnames)
otu_prev_filt_list <- otu_rel %>% 
  map2(prevalent_otus_list,~.x[rownames(.x) %in% .y])
#Next filter bugs by abundance - mean abundance of at least 0.0005%
otu_prev_abund_list <- otu_prev_filt_list %>% 
  map(~.[rowMeans(.)>0.000005])
#Organize into matrix form and print
otu_mat <- otu_prev_abund_list %>% 
  map(as.matrix) %>% 
  map(t)
#walk2(otu_mat, c("Species_otu","Genus_otu", "Family_otu"),
 #     ~write.csv(.x, here::here(str_c("corr_mats/",.y,".csv", sep = ""))))
```

Create clr-transformed matrices for 16S data (keep all the previous filters)

```{r clr-16s}
#Filter the otu tables with raw counts for prevalence and abundance
otu_filt_list <- list(physeq_intersect,phy_genus, phy_family) %>% 
  map(otu_table) %>% 
  map2(otu_mat, ~ .x [rownames(.x) %in% colnames(.y)])
#Apply clr transformation to the otu tables
otu_filt_clr <- map(otu_filt_list, qiime2R::make_clr)
#Check how the data look after transformation
otu_filt_clr %>% 
  map(rowMeans) %>% 
  walk(hist)
```


```{r clr-16s-print, eval=FALSE}
otu_filt_clr %>% 
map(as.matrix) %>% 
  map(t) %>% 
walk2(c("Species_otu","Genus_otu", "Family_otu"),
      ~write.csv(.x, here::here(str_c("corr_mats/clr",.y,".csv", sep = ""))))
                      
```

Prepare a table of picrust-pathways for correlation
```{r picrust-pathway}
pth <- read_tsv(here::here("picrust_res/pathway-table.tsv"), skip = 1)
pth_mat <- pth %>%  
  column_to_rownames(var = "#OTU ID") %>% 
  as.matrix() 
colnames(pth_mat) <- str_remove(colnames(pth_mat),"Rep")
pth_mat <- pth_mat [,rownames(mq_mat)] 
#Remove low prevalence pathways - expect prevalence in at least 30%
pth_prevalence_filter <- pth_mat %>%
  t() %>% 
  as.data.frame() %>% 
  mutate(across(everything(), ~.>0))
pth_prev_sum <- pth_prevalence_filter %>% 
  summarise(across(everything(), sum)) %>% 
  mutate(across(everything(),~./ncol(pth_mat)))
prevalent_pth <- pth_prev_sum %>% 
  select(where(~.x >= 0.3)) %>% 
  colnames()
pth_prev <- pth_mat [prevalent_pth,]
pth_clr <- qiime2R::make_clr(pth_prev)


pth_mat %>% t %>% rowMeans() %>% hist(main = "Untransformed pathways")
pth_prev %>% t %>% rowMeans() %>% hist(main = "Prevalent pathways")
pth_clr %>% t %>% rowMeans() %>% hist(main = "Prevalent clr pathways")
#write.csv(t(pth_clr), here::here("corr_mats/clr_pth.csv"))
```





